compiler/hash-tir/src/environment/stores.rs

```diff
@@ -66,33 +85,265 @@ macro_rules! tir_debug_value_of_sequence_store_element_id {
+                use hash_storage::store::statics::CoreStoreId;
                 f.debug_tuple(stringify!($id)).field(&self.index).field(&self.value()).finish()
             }
         }
     };
 }
 
+/// Debug any store element by only printing its name (element values must have
+/// a field `.main`).
 #[macro_export]
 macro_rules! tir_debug_name_of_store_id {
     ($id:ident) => {
         impl std::fmt::Debug for $id {
             fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-                use hash_storage::store::statics::StoreId;
+                use hash_storage::store::statics::CoreStoreId;
                 f.debug_tuple(stringify!($id)).field(&self.value().name).finish()
             }
         }
     };
 }
 
+/// Define a TIR node that is stored in a single store.
+///
+/// This internally uses `hash_storage::static_single_store!` to define the
+/// store and its ID type; it creates a single store with values
+/// `Node<$value>`.
+///
+/// This can be used to define TIR nodes like:
+///
+/// ```ignore
+/// tir_node_single_store!(Term);
+///
+/// /// These now exist:
+/// pub type TermStore: Store<TermId, Node<Term>>;
+/// pub type TermId: CoreStoreId<Value=Term>;
+/// pub fn term_store() -> &'static TermStore;
+/// ```
+#[macro_export]
+macro_rules! tir_node_single_store {
+    // Generate the other names from the given `$name`:
+    ($name:ident) => {
+        paste::paste! {
+            tir_node_single_store!(
+                store = pub [<$name:camel Store>],
+                id = pub [<$name:camel Id>],
+                value = $name,
+                store_name = [<$name:snake>]
+            );
+        }
+    };
+    (
+        store = $store_vis:vis $store:ident,
+        id = $id_vis:vis $id:ident,
+        value = $value:ty,
+        store_name = $store_name:ident
+    ) => {
+        hash_storage::static_single_store!(
+            store = $store_vis $store,
+            id = $id_vis $id,
+            value = $crate::node::Node<$value>,
+            store_name = $store_name,
+            store_source = tir_stores()
+        );
+
+        $crate::tir_debug_value_of_single_store_id!($id);
+
+        // @@Todo: enable once locations are properly set up
+        // impl $crate::ast_info::HasNodeId for $id {
+        //     fn node_id(&self) -> Option<hash_ast::ast::AstNodeId> {
+        //         self.value().node_id()
+        //     }
+        // }
+    };
+}
+
+/// Define a TIR node that is stored in a direct sequence store.
+///
+/// This internally uses `hash_storage::static_single_store!` and
+/// `hash_storage::static_sequence_store_direct!` to define the store and its ID
+/// type; it creates a single store from key `$id` to value `Node<$id_seq>`, and
+/// a direct sequence store from key `$id_seq` to values in `$value`.
+///
+/// This can be used to define TIR nodes like:
+///
+/// ```ignore
+/// tir_node_sequence_store_direct!(Param);
+///
+/// /// These now exist:
```

Comment:
Maybe add docs for what each one is responsible for and its purpose? Specifically `ParamsSeqId` and `ParamsId`